File: lec11/02_dfs/dfs.py
"""
DFS (Depth-First Search) 알고리즘 구현
그래프에서 깊이 우선 탐색을 수행합니다.

시간 복잡도: O(V + E) - V는 정점 수, E는 간선 수
공간 복잡도: O(V) - 스택과 방문 집합을 위한 공간
"""

from collections import defaultdict
from typing import List, Set


class Graph:
  """
  그래프 클래스 (인접 리스트로 표현)
  """

  def __init__(self, num_vertices: int = 0):
    """
    그래프 초기화

    Args:
        num_vertices: 정점의 개수 (옵션)
    """
    self.num_vertices = num_vertices
    self.adj_list = defaultdict(list)

  def add_edge(self, src: int, dest: int) -> None:
    """
    간선 추가 (무방향 그래프)

    Args:
        src: 시작 정점
        dest: 도착 정점
    """
    self.adj_list[src].append(dest)
    self.adj_list[dest].append(src)  # 무방향 그래프

  def add_directed_edge(self, src: int, dest: int) -> None:
    """
    방향 간선 추가 (방향 그래프)

    Args:
        src: 시작 정점
        dest: 도착 정점
    """
    self.adj_list[src].append(dest)

  def get_vertices(self) -> Set[int]:
    """
    그래프의 모든 정점 반환

    Returns:
        모든 정점의 집합
    """
    vertices = set()
    for vertex in self.adj_list:
      vertices.add(vertex)
      vertices.update(self.adj_list[vertex])
    return vertices

  def print_graph(self) -> None:
    """
    그래프 구조 출력 (디버깅용)
    """
    print("\nGraph representation (adjacency list):")
    vertices = sorted(self.get_vertices())
    for vertex in vertices:
      neighbors = self.adj_list[vertex]
      print(f"Vertex {vertex}: {neighbors}")
    print()


def dfs_recursive(graph: Graph, start_vertex: int) -> List[int]:
  """
  DFS 알고리즘 구현 (재귀적)

  Args:
      graph: 탐색할 그래프
      start_vertex: 시작 정점

  Returns:
      DFS 탐색 순서의 정점 리스트
  """
  if start_vertex not in graph.get_vertices():
    print(f"Invalid start vertex: {start_vertex}")
    return []

  # 방문 상태를 추적하는 집합
  explored = set()

  # DFS 탐색 결과를 저장할 리스트
  dfs_order = []

  def dfs_visit(vertex: int) -> None:
    """
    DFS 재귀 보조 함수

    Args:
        vertex: 현재 방문할 정점
    """
    # 현재 정점을 방문 처리
    explored.add(vertex)
    dfs_order.append(vertex)

    # 현재 정점의 모든 인접 정점에 대해 재귀 호출
    # 정렬해서 일관된 순서로 탐색
    neighbors = sorted(graph.adj_list[vertex])
    for neighbor in neighbors:
      if neighbor not in explored:
        dfs_visit(neighbor)

  print(f"DFS Recursive Traversal starting from vertex {start_vertex}: ", end="")

  # 재귀 DFS 시작
  dfs_visit(start_vertex)

  # 결과 출력
  print(" ".join(map(str, dfs_order)))
  return dfs_order


def dfs_iterative(graph: Graph, start_vertex: int) -> List[int]:
  """
  DFS 알고리즘 구현 (반복적) - 스택 상태 시각화 포함

  Args:
      graph: 탐색할 그래프
      start_vertex: 시작 정점

  Returns:
      DFS 탐색 순서의 정점 리스트
  """
  if start_vertex not in graph.get_vertices():
    print(f"Invalid start vertex: {start_vertex}")
    return []

  # 방문 상태를 추적하는 집합
  explored = set()

  # 스택 초기화 (리스트 사용)
  stack = [start_vertex]
  print(f"Initial stack: {stack}")

  # DFS 탐색 결과를 저장할 리스트
  dfs_order = []

  print(f"DFS Iterative Traversal starting from vertex {start_vertex}:")

  # DFS 메인 루프
  while stack:
    print(f"  Stack before pop: {stack}", end="")

    # 스택에서 정점을 하나 꺼냄
    current_vertex = stack.pop()
    print(f" -> Pop {current_vertex} -> Stack after pop: {stack}")

    # 아직 탐색되지 않은 정점이라면
    if current_vertex not in explored:
      explored.add(current_vertex)
      dfs_order.append(current_vertex)
      print(f"  Visit: {current_vertex}")

      # 현재 정점의 모든 인접 정점을 스택에 추가 (역순으로)
      # 일관된 순서를 위해 정렬 후 역순으로 추가
      neighbors = sorted(graph.adj_list[current_vertex], reverse=True)
      for neighbor in neighbors:
        if neighbor not in explored:
          stack.append(neighbor)
          print(f"  Push {neighbor} -> Stack: {stack}")
    else:
      print(f"  {current_vertex} already visited, skip")

    print()  # 빈 줄

  print(f"Final traversal order: {' '.join(map(str, dfs_order))}")
  return dfs_order


def dfs_find_path(graph: Graph, start: int, target: int) -> List[int]:
  """
  DFS를 이용한 경로 찾기

  Args:
      graph: 탐색할 그래프
      start: 시작 정점
      target: 목표 정점

  Returns:
      경로를 나타내는 정점 리스트 (경로가 없으면 빈 리스트)
  """
  if start not in graph.get_vertices() or target not in graph.get_vertices():
    return []

  if start == target:
    return [start]

  explored = set()
  parent = {}

  def dfs_visit(vertex: int) -> bool:
    """
    DFS로 목표 정점을 찾는 재귀 함수

    Args:
        vertex: 현재 정점

    Returns:
        목표 정점을 찾았으면 True, 아니면 False
    """
    explored.add(vertex)

    if vertex == target:
      return True

    neighbors = graph.adj_list[vertex]
    for neighbor in neighbors:
      if neighbor not in explored:
        parent[neighbor] = vertex
        if dfs_visit(neighbor):
          return True

    return False

  # DFS로 경로 찾기
  if dfs_visit(start):
    # 경로 재구성
    path = []
    current = target
    while current is not None:
      path.append(current)
      current = parent.get(current)
    return path[::-1]  # 역순으로 반환

  return []  # 경로가 없음


def main():
  """
  메인 함수 - 예제 실행
  """
  print("=== DFS (Depth-First Search) Algorithm ===\n")

  # 그래프 생성 (7개 정점)
  graph = Graph()

  # 간선 추가 (C 코드와 동일한 그래프)
  graph.add_edge(0, 1)
  graph.add_edge(0, 2)
  graph.add_edge(0, 5)
  graph.add_edge(0, 6)
  graph.add_edge(3, 4)
  graph.add_edge(3, 5)
  graph.add_edge(4, 5)
  graph.add_edge(4, 6)

  # 그래프 구조 출력
  graph.print_graph()

  # DFS 실행 (한 개 예제만)
  print("=== DFS Recursive ===")
  dfs_recursive(graph, 0)

  print("\n=== DFS Iterative (with stack visualization) ===")
  dfs_iterative(graph, 0)

  print("\n=== DFS with different graph ===")

  # 다른 그래프 예제
  graph2 = Graph()
  graph2.add_edge(0, 1)
  graph2.add_edge(0, 4)
  graph2.add_edge(1, 2)
  graph2.add_edge(1, 3)
  graph2.add_edge(1, 4)
  graph2.add_edge(2, 3)
  graph2.add_edge(3, 4)

  graph2.print_graph()
  print("=== DFS Recursive ===")
  dfs_recursive(graph2, 0)
  print("\n=== DFS Iterative (with stack visualization) ===")
  dfs_iterative(graph2, 0)


if __name__ == "__main__":
  main()

File: lec11/02_dfs/dfs.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

#define MAX_VERTICES 100

/**
 * 스택 구조체와 관련 함수들
 */
typedef struct {
  int data[MAX_VERTICES];
  int top;
} Stack;

// 스택 초기화
void initStack(Stack* s) {
  s->top = -1;
}

// 스택이 비어있는지 확인
bool isStackEmpty(Stack* s) {
  return s->top == -1;
}

// 스택이 가득 찼는지 확인
bool isStackFull(Stack* s) {
  return s->top == MAX_VERTICES - 1;
}

// 스택에 원소 추가 (push)
void push(Stack* s, int value) {
  if (isStackFull(s)) {
    printf("Stack is full!\n");
    return;
  }
  s->data[++s->top] = value;
}

// 스택 상태 출력 (디버깅용)
void printStack(Stack* s) {
  printf("[");
  for (int i = 0; i <= s->top; i++) {
    printf("%d", s->data[i]);
    if (i < s->top)
      printf(", ");
  }
  printf("]");
}

// 스택에서 원소 제거 (pop)
int pop(Stack* s) {
  if (isStackEmpty(s)) {
    printf("Stack is empty!\n");
    return -1;
  }
  return s->data[s->top--];
}

/**
 * 그래프 구조체 (인접 리스트로 표현)
 */
typedef struct Node {
  int vertex;
  struct Node* next;
} Node;

typedef struct {
  int numVertices;
  Node** adjList;
} Graph;

// 그래프 생성
Graph* createGraph(int vertices) {
  Graph* graph = (Graph*)malloc(sizeof(Graph));
  graph->numVertices = vertices;
  graph->adjList = (Node**)malloc(vertices * sizeof(Node*));

  for (int i = 0; i < vertices; i++) {
    graph->adjList[i] = NULL;
  }

  return graph;
}

// 간선 추가 (무방향 그래프)
void addEdge(Graph* graph, int src, int dest) {
  // src -> dest 간선 추가
  Node* newNode = (Node*)malloc(sizeof(Node));
  newNode->vertex = dest;
  newNode->next = graph->adjList[src];
  graph->adjList[src] = newNode;

  // dest -> src 간선 추가 (무방향 그래프)
  newNode = (Node*)malloc(sizeof(Node));
  newNode->vertex = src;
  newNode->next = graph->adjList[dest];
  graph->adjList[dest] = newNode;
}

/**
 * DFS 알고리즘 구현 (반복적)
 * @param graph: 탐색할 그래프
 * @param startVertex: 시작 정점
 */
void DFS_Iterative(Graph* graph, int startVertex) {
  if (startVertex >= graph->numVertices || startVertex < 0) {
    printf("Invalid start vertex!\n");
    return;
  }

  // 방문 상태를 추적하는 배열
  bool* explored = (bool*)calloc(graph->numVertices, sizeof(bool));

  // 스택 초기화
  Stack s;
  initStack(&s);

  // 시작 정점을 스택에 추가
  push(&s, startVertex);
  printf("Initial stack: ");
  printStack(&s);
  printf("\n");

  printf("DFS Iterative Traversal starting from vertex %d:\n", startVertex);

  // DFS 메인 루프
  while (!isStackEmpty(&s)) {
    printf("  Stack before pop: ");
    printStack(&s);

    // 스택에서 정점을 하나 꺼냄
    int currentVertex = pop(&s);
    printf(" -> Pop %d -> Stack after pop: ", currentVertex);
    printStack(&s);
    printf("\n");

    // 아직 탐색되지 않은 정점이라면
    if (!explored[currentVertex]) {
      explored[currentVertex] = true;
      printf("  Visit: %d\n", currentVertex);

      // 현재 정점의 모든 인접 정점을 스택에 추가 (역순으로)
      // 일관된 순서를 위해 배열에 저장 후 역순으로 push
      int neighbors[MAX_VERTICES];
      int neighborCount = 0;

      Node* temp = graph->adjList[currentVertex];
      while (temp != NULL) {
        neighbors[neighborCount++] = temp->vertex;
        temp = temp->next;
      }

      // 역순으로 스택에 추가
      for (int i = neighborCount - 1; i >= 0; i--) {
        if (!explored[neighbors[i]]) {
          push(&s, neighbors[i]);
          printf("  Push %d -> Stack: ", neighbors[i]);
          printStack(&s);
          printf("\n");
        }
      }
    } else {
      printf("  %d already visited, skip\n", currentVertex);
    }
    printf("\n");
  }

  printf("\n");
  free(explored);
}

/**
 * DFS 알고리즘 구현 (재귀적) - 보조 함수
 */
void DFS_Visit(Graph* graph, int vertex, bool* explored) {
  // 현재 정점을 방문 처리
  explored[vertex] = true;
  printf("%d ", vertex);

  // 현재 정점의 모든 인접 정점에 대해 재귀 호출
  Node* temp = graph->adjList[vertex];
  while (temp != NULL) {
    int adjVertex = temp->vertex;
    if (!explored[adjVertex]) {
      DFS_Visit(graph, adjVertex, explored);
    }
    temp = temp->next;
  }
}

/**
 * DFS 알고리즘 구현 (재귀적)
 * @param graph: 탐색할 그래프
 * @param startVertex: 시작 정점
 */
void DFS_Recursive(Graph* graph, int startVertex) {
  if (startVertex >= graph->numVertices || startVertex < 0) {
    printf("Invalid start vertex!\n");
    return;
  }

  // 방문 상태를 추적하는 배열
  bool* explored = (bool*)calloc(graph->numVertices, sizeof(bool));

  printf("DFS Recursive Traversal starting from vertex %d: ", startVertex);

  // 재귀 DFS 시작
  DFS_Visit(graph, startVertex, explored);

  printf("\n");
  free(explored);
}

// 그래프 출력 (디버깅용)
void printGraph(Graph* graph) {
  printf("\nGraph representation (adjacency list):\n");
  for (int i = 0; i < graph->numVertices; i++) {
    printf("Vertex %d: ", i);
    Node* temp = graph->adjList[i];
    while (temp != NULL) {
      printf("%d -> ", temp->vertex);
      temp = temp->next;
    }
    printf("NULL\n");
  }
  printf("\n");
}

// 그래프 메모리 해제
void freeGraph(Graph* graph) {
  for (int i = 0; i < graph->numVertices; i++) {
    Node* temp = graph->adjList[i];
    while (temp != NULL) {
      Node* toDelete = temp;
      temp = temp->next;
      free(toDelete);
    }
  }
  free(graph->adjList);
  free(graph);
}

/**
 * 메인 함수 - 예제 실행
 */
int main() {
  printf("=== DFS (Depth-First Search) Algorithm ===\n\n");

  // 그래프 생성 (7개 정점)
  Graph* graph = createGraph(7);

  // 간선 추가
  //   addEdge(graph, 0, 1);
  //   addEdge(graph, 0, 2);
  //   addEdge(graph, 1, 3);
  //   addEdge(graph, 1, 4);
  //   addEdge(graph, 2, 5);
  //   addEdge(graph, 2, 6);

  addEdge(graph, 0, 1);
  addEdge(graph, 0, 2);
  addEdge(graph, 0, 5);
  addEdge(graph, 0, 6);
  addEdge(graph, 3, 4);
  addEdge(graph, 3, 5);
  addEdge(graph, 4, 5);
  addEdge(graph, 4, 6);

  // 그래프 구조 출력
  printGraph(graph);

  // DFS 실행 (한 개 예제만)
  printf("=== DFS Recursive ===\n");
  DFS_Recursive(graph, 0);

  printf("\n=== DFS Iterative (with stack visualization) ===\n");
  DFS_Iterative(graph, 0);

  printf("\n=== DFS with different graph ===\n");

  // 다른 그래프 예제
  Graph* graph2 = createGraph(5);
  addEdge(graph2, 0, 1);
  addEdge(graph2, 0, 4);
  addEdge(graph2, 1, 2);
  addEdge(graph2, 1, 3);
  addEdge(graph2, 1, 4);
  addEdge(graph2, 2, 3);
  addEdge(graph2, 3, 4);

  printGraph(graph2);
  DFS_Recursive(graph2, 0);
  DFS_Iterative(graph2, 0);

  // 메모리 해제
  freeGraph(graph);
  freeGraph(graph2);

  return 0;
}
